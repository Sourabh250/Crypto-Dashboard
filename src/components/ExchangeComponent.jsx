import React, { useState } from "react";
import { IoTriangle } from "react-icons/io5";

function ExchangeComponent() {
  const [sellingCurrency, setSellingCurrency] = useState("Btc");
  const [buyingCurrency, setBuyingCurrency] = useState("Eth");
  const [sellList, setSellList] = useState(false);
  const [buyList, setBuyList] = useState(false);
  const [sellAmount, setSellAmount] = useState("");
  const [buyAmount, setBuyAmount] = useState(null);
  const [error, setError] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  // List of currencies
  const currencyList = [
    "Btc",
    "Eth",
    "Xrp",
    "Bnb",
    "Dot",
    "Ltc",
    "Usd",
    "Eur",
    "Gbp",
    "Inr",
    "Jpy",
  ];

  // Handler for selecting the currency to sell
  const handleSellCurrency = (e) => {
    setSellingCurrency(e.target.textContent);
    setSellList(false);
  };
  
  // Handler for selecting the currency to buy
  const handleBuyCurrency = (e) => {
    setBuyingCurrency(e.target.textContent);
    setBuyList(false);
  };

  // Handler for the exchange
  const handleExchange = async () => {
    setError(""); // Reset error message

    // Validate the sell amount
    if(isNaN(sellAmount) || sellAmount <= 0) {
        setError("Invalid Number");
        return;
    }

    try {
      setIsLoading(true);
      // Fetching exchange rates
      const response = await fetch(
        `https://api.coingecko.com/api/v3/exchange_rates`
      );
      if (!response.ok) {
        throw new Error('Failed to fetch. Please try again later.');
      }
      const data = await response.json();
      const rates = data.rates;

      // Calculating the conversion rate
      const sellRate = rates[sellingCurrency.toLowerCase()].value;
      const buyRate = rates[buyingCurrency.toLowerCase()].value;
      const conversionRate = buyRate / sellRate;
      const amountConverted = (sellAmount * conversionRate).toFixed(2);
      setBuyAmount(amountConverted); // Setting the converted buy amount
      setIsLoading(false);
    } catch (error) {
      console.error(error);
      setError("Limit exceeded"); // If fetch fails, displaying error
    }
  };

  return (
    <div className="bg-gray-100 rounded-lg shadow-lg p-4">
      <h1 className="text-xl font-semibold">Exchange Coins</h1>
      <div className="flex gap-4 lg:gap-8 my-6 items-center justify-center sm:justify-normal">
        <h1 className="text-red-600">Sell</h1>
        <div className="w-36 relative">

          {/* Dropdown for selecting the currency to sell  */}
          <button
            onClick={() => setSellList(!sellList)}
            className="w-full bg-gray-200 px-4 py-2 rounded flex justify-between items-center gap-2"
          >
            {sellingCurrency}
            <IoTriangle style={{ transform: "rotate(180deg)" }} />
          </button>
          {sellList && (
            <ul data-testid="sell-currency-list" className="absolute z-20 mt-2 rounded shadow-md bg-gray-300 p-2 max-h-20 overflow-y-auto w-36  custom-scrollbar">
              {currencyList.map((currency) => (
                <li
                  key={currency}
                  onClick={handleSellCurrency}
                  className="p-1 hover:bg-gray-200 cursor-pointer"
                >
                  {currency}
                </li>
              ))}
            </ul>
          )}
        </div>
        {/* Input for entering the sell amount */}
        <div className="relative w-36">
          <label htmlFor="sell-value" className="block text-sm font-medium text-gray-700 absolute -top-6">
            Enter value
          </label>
          <input id="sell-value"
            onChange={(e) => setSellAmount(e.target.value)}
            className="w-full px-4 py-2 rounded"
          />
          {/* Error message if sell amount is invalid */}
          {error && <small className="absolute block -bottom-6 text-red-500">{error}</small>}
        </div>
      </div>
      <div className="flex gap-4 lg:gap-8 items-center justify-center sm:justify-normal">
        <h1 className="text-green-600">Buy</h1>
        <div className="w-36 relative">
          {/* Dropdown for selecting the currency to buy */}
          <button
            onClick={() => setBuyList(!buyList)}
            className="w-full bg-gray-200 px-4 py-2 rounded flex justify-between items-center gap-2"
          >
            {buyingCurrency}
            <IoTriangle style={{ transform: "rotate(180deg)" }} />
          </button>
          {buyList && (
            <ul data-testid="buy-currency-list" className="absolute z-10 mt-2 rounded shadow-md bg-gray-300 p-2 max-h-20 overflow-y-auto w-36 custom-scrollbar">
              {currencyList.map((currency) => (
                <li
                  key={currency}
                  onClick={handleBuyCurrency}
                  className="p-1 hover:bg-gray-200 cursor-pointer"
                >
                  {currency}
                </li>
              ))}
            </ul>
          )}
        </div>
        {/* Display the converted buy amount */}
        {isLoading ? (
          <div className="w-36 flex justify-center items-center">
          <div className="spinner"></div>
        </div>
        ) : (
          <div className="w-36 text-center text-green-600">{buyAmount}</div>
        )}
      </div>
      {/* Button to initiate the exchange */}
      <div className="mt-6 flex justify-center">
        <button
          className="bg-blue-600 px-4 py-2 rounded-md text-gray-200"
          onClick={handleExchange}
        >
          Exchange
        </button>
      </div>
    </div>
  );
}

export default ExchangeComponent;