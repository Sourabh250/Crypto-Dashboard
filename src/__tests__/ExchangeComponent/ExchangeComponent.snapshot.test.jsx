import React from "react";
import renderer from 'react-test-renderer';
import  ExchangeComponent from '../../components/ExchangeComponent';

test('ExchangeComponent snapshot', () => {
    const tree = renderer.create(<ExchangeComponent />).toJSON();

    expect(tree).toMatchSnapshot();
});