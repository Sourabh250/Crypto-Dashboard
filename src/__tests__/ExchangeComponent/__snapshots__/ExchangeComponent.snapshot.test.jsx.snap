// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeComponent snapshot 1`] = `
<div
  className="bg-gray-100 rounded-lg shadow-lg p-4"
>
  <h1
    className="text-xl font-semibold"
  >
    Exchange Coins
  </h1>
  <div
    className="flex gap-4 lg:gap-8 my-6 items-center justify-center sm:justify-normal"
  >
    <h1
      className="text-red-600"
    >
      Sell
    </h1>
    <div
      className="w-36 relative"
    >
      <button
        className="w-full bg-gray-200 px-4 py-2 rounded flex justify-between items-center gap-2"
        onClick={[Function]}
      >
        Btc
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
              "transform": "rotate(180deg)",
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 464H48a16 16 0 0 1-14.07-23.62l208-384a16 16 0 0 1 28.14 0l208 384A16 16 0 0 1 464 464z"
          />
        </svg>
      </button>
    </div>
    <div
      className="relative w-36"
    >
      <label
        className="block text-sm font-medium text-gray-700 absolute -top-6"
        htmlFor="sell-value"
      >
        Enter value
      </label>
      <input
        className="w-full px-4 py-2 rounded"
        id="sell-value"
        onChange={[Function]}
      />
    </div>
  </div>
  <div
    className="flex gap-4 lg:gap-8 items-center justify-center sm:justify-normal"
  >
    <h1
      className="text-green-600"
    >
      Buy
    </h1>
    <div
      className="w-36 relative"
    >
      <button
        className="w-full bg-gray-200 px-4 py-2 rounded flex justify-between items-center gap-2"
        onClick={[Function]}
      >
        Eth
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
              "transform": "rotate(180deg)",
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 464H48a16 16 0 0 1-14.07-23.62l208-384a16 16 0 0 1 28.14 0l208 384A16 16 0 0 1 464 464z"
          />
        </svg>
      </button>
    </div>
    <div
      className="w-36 text-center text-green-600"
    />
  </div>
  <div
    className="mt-6 flex justify-center"
  >
    <button
      className="bg-blue-600 px-4 py-2 rounded-md text-gray-200"
      onClick={[Function]}
    >
      Exchange
    </button>
  </div>
</div>
`;
